analyses/cms-open-data-ttbar/analysis.py

```diff
@@ -147,62 +147,43 @@ def define_trijet_mass(df: ROOT.RDataFrame) -> ROOT.RDataFrame:
 
     # Build trijet combinations
     df = df.Define("Trijet", "ROOT::VecOps::Combinations(Jet_pt[Jet_pt_mask],3)")
-    df = df.Define("nTrijet", "Trijet[0].size()")
 
-    # Assign four-momentums to each trijet combination
+    
+    # trijet_btag is a helpful array of bool values indicating whether or not the maximum btag value in trijet is larger than 0.5 threshold
     df = df.Define(
-        "Trijet_p4",
-        """
-        ROOT::RVec<ROOT::Math::PxPyPzMVector> Trijet_p4(nTrijet);
-        for (int i = 0; i < nTrijet; ++i)
-        {
-            int j1 = Trijet[0][i];
-            int j2 = Trijet[1][i];
-            int j3 = Trijet[2][i];
-            Trijet_p4[i] = Jet_p4[j1] + Jet_p4[j2] + Jet_p4[j3];
-        }
-        return Trijet_p4;
-        """,
+            "Trijet_btag",
+            """
+            auto Jet_btagCSVV2_masked = Jet_btagCSVV2[Jet_pt_mask];
+            auto J1_btagCSVV2 = ROOT::VecOps::Take(Jet_btagCSVV2_masked, Trijet[0]);
+            auto J2_btagCSVV2 = ROOT::VecOps::Take(Jet_btagCSVV2_masked, Trijet[1]);
+            auto J3_btagCSVV2 = ROOT::VecOps::Take(Jet_btagCSVV2_masked, Trijet[2]);
+            return (J1_btagCSVV2>0.5)||(J2_btagCSVV2>0.5)||(J3_btagCSVV2>0.5);
+            """
+            # FIXME 
+            # Do insteam something like max(J1_btag,J2_btag, Jt3_btag)>0.5. 
+            # Do I need to define custom function max(RVec 1, RVec2, RVec3)?
+        )
+
+    # Assign four-momentums to each trijet combination
+    df = (
+        df.Define('J1', 'ROOT::VecOps::Take(Jet_p4, Trijet[0])').
+        Define('J2', 'ROOT::VecOps::Take(Jet_p4, Trijet[1])').
+        Define('J3', 'ROOT::VecOps::Take(Jet_p4, Trijet[2])').
+        Define('Trijet_p4', '(J1+J2+J3)[Trijet_btag]')
```

Comment:
I agree this is more readable than the for loop, however 4 Defines are slower than 1 :smile: if we don't need J1, J2 and J3 anywhere else we can just create local variables like you do for `Trijet_btag` I guess?

Reply:
Yes, I've defined `Jn` as separate columns as I needed them for debugging numerical discrepancies. Thank you for pointing it out. I removed them